src/AMSlib/wf/resource_manager.cpp

```diff
@@ -5,48 +5,113 @@
  * SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
  */
 
-#include "resource_manager.hpp"
-
+#include <cstdlib>
+#include <cstring>
 #include <umpire/ResourceManager.hpp>
 #include <umpire/Umpire.hpp>
 #include <umpire/strategy/QuickPool.hpp>
 
 #include "debug.h"
+#include "device.hpp"
+#include "resource_manager.hpp"
 
 namespace ams
 {
 
-std::string AMSAllocator::getName() { return allocator.getName(); }
-
-void *AMSAllocator::allocate(size_t num_bytes)
+template<typename T>
+static T roundUp(T num_to_round, int multiple) 
 {
-  void *ptr = allocator.allocate(num_bytes);
-  CFATAL(ResourceManager,
-         ptr == nullptr,
-         "Failed to allocated %ld values using allocator %s",
-         num_bytes,
-         getName().c_str());
-  return ptr;
+    return ((num_to_round + multiple- 1) / multiple) * multiple;
 }
 
-void AMSAllocator::deallocate(void *ptr) { allocator.deallocate(ptr); }
+std::string AMSAllocator::getName() { return name; }
+
+
+struct AMSDefaultDeviceAllocator final : AMSAllocator {
+  AMSDefaultDeviceAllocator(std::string name) : AMSAllocator(name){};
+  ~AMSDefaultDeviceAllocator() = default;
+
+  void *allocate(size_t num_bytes) { return DeviceAllocate(num_bytes); }
+
+  void deallocate(void *ptr) { return DeviceFree(ptr); }
+};
+
+struct AMSDefaultHostAllocator final : AMSAllocator {
+  AMSDefaultHostAllocator(std::string name) : AMSAllocator(name) {}
+  ~AMSDefaultHostAllocator() = default;
 
-void AMSAllocator::registerPtr(void *ptr, size_t nBytes)
+  void *allocate(size_t num_bytes) { return aligned_alloc(8, roundUp(num_bytes, 8)); }
+
+  void deallocate(void *ptr) { free(ptr); }
+};
+
+struct AMSDefaultPinnedAllocator final : AMSAllocator {
+  AMSDefaultPinnedAllocator(std::string name) : AMSAllocator(name) {}
+  ~AMSDefaultPinnedAllocator() = default;
+
+  void *allocate(size_t num_bytes) { return DevicePinnedAlloc(num_bytes); }
+
+  void deallocate(void *ptr) { DeviceFreePinned(ptr); }
+};
+
+
+namespace internal
 {
-  auto &rm = umpire::ResourceManager::getInstance();
-  rm.registerAllocation(ptr,
-                        umpire::util::AllocationRecord(
-                            ptr, nBytes, allocator.getAllocationStrategy()));
+void _raw_copy(void *src,
+               AMSResourceType src_dev,
+               void *dest,
+               AMSResourceType dest_dev,
+               size_t num_bytes)
+{
+  if (src_dev == AMSResourceType::AMS_HOST) {
```

Comment:
nested switch-case with fall-throughs to remove duplication for readability?

Reply:
Fixed.